core/src/ephemeral_server/mod.rs

```diff
@@ -0,0 +1,461 @@
+                    other => return Err(anyhow!("Unsupported arch: {}", other)),
+                };
+                let mut get_info_params = vec![("arch", arch), ("platform", platform)];
+                let version_name = match version {
+                    EphemeralExeVersion::Default {
+                        sdk_name,
+                        sdk_version,
+                    } => {
+                        get_info_params.push(("sdk-name", sdk_name.as_str()));
+                        get_info_params.push(("sdk-version", sdk_version.as_str()));
+                        "default"
+                    }
+                    EphemeralExeVersion::Fixed(version) => version,
+                };
+                let client = reqwest::Client::new();
+                let info: DownloadInfo = client
+                    .get(format!(
+                        "https://temporal.download/{}/{}",
+                        artifact_name, version_name
+                    ))
+                    .query(&get_info_params)
+                    .send()
+                    .await?
+                    .json()
+                    .await?;
+
+                // Attempt download, looping because it could have waited for
+                // concurrent one to finish
+                loop {
+                    if lazy_download_exe(
+                        &client,
+                        &info.archive_url,
+                        Path::new(&info.file_to_extract),
+                        &dest,
+                    )
+                    .await?
+                    {
+                        return Ok(dest);
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn get_free_port() -> u16 {
+    // Can just ask OS to give us a port then close socket. OS's don't give that
+    // port back to anyone else anytime soon.
+    std::net::TcpListener::bind("127.0.0.1:0")
+        .unwrap()
+        .local_addr()
+        .unwrap()
+        .port()
+}
+
+// Returns false if we successfully waited for another download to complete, or
+// true if the destination is known to exist. Should call again if false is
+// returned.
+async fn lazy_download_exe(
+    client: &reqwest::Client,
+    uri: &str,
+    file_to_extract: &Path,
+    dest: &Path,
+) -> anyhow::Result<bool> {
+    // If it already exists, do not extract
+    if dest.exists() {
+        return Ok(true);
+    }
+
+    // We only want to download if we're not already downloading. To avoid some
+    // kind of global lock, we'll just create the file eagerly w/ a temp
+    // filename and delete it on failure or move it on success. If the temp file
+    // already exists, we'll wait a bit and re-run this.
+    let temp_dest_str = format!("{}{}", dest.to_str().unwrap(), ".downloading");
+    let temp_dest = Path::new(&temp_dest_str);
+    // This match only gets Ok if the file was downloaded and extracted to the
+    // temporary path
+    match new_executable_file(temp_dest) {
```

Comment:
I would actually inline this function since the behavior here depends heavily on `create_new` being set true and it's more obvious when that's right here.

You can use the `cfg` guards on different `let` statements binding the same variable name

Reply:
The reason I don't inline this is because it has different behavior per OS (I set it as executable for unix, doesn't apply for windows)

Reply:
Done